# **EDA:**
## Exploratory Data Analysis  

Exploratory Data Analysis (EDA) in Python is the foundational process of summarizing, cleaning, and visualizing a dataset to uncover patterns, detect anomalies, and test underlying assumptions. It is a critical initial step in data science before diving into heavy statistical modeling or machine learning.

In [7]:
# importing libraries

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


# 1. import dataset

In [8]:
# importing data

df = sns.load_dataset('titanic')
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


# 2. Bigger picture of the data

In [29]:
df.dtypes

survived          int64
pclass            int64
sex                 str
age             float64
sibsp             int64
parch             int64
fare            float64
embarked            str
class          category
who                 str
adult_male         bool
deck           category
embark_town         str
alive               str
alone              bool
dtype: object

In [30]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          891 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


In [31]:
# Finding missing values (null_values)

df.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [32]:
df

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.000000,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.000000,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.000000,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [33]:
df2 = df

In [34]:
df2

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.000000,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.000000,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.000000,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [35]:
# Calculating missing values in percentage

(df2.isnull().sum() / len(df2)) * 100

survived        0.000000
pclass          0.000000
sex             0.000000
age             0.000000
sibsp           0.000000
parch           0.000000
fare            0.000000
embarked        0.224467
class           0.000000
who             0.000000
adult_male      0.000000
deck           77.216611
embark_town     0.224467
alive           0.000000
alone           0.000000
dtype: float64

In [36]:
print(df2['age'].max())
print(df2['age'].min())
print(df2['age'].mean())

80.0
0.42
29.69911764705882


In [37]:
df2['age'].unique()

array([22.        , 38.        , 26.        , 35.        , 29.69911765,
       54.        ,  2.        , 27.        , 14.        ,  4.        ,
       58.        , 20.        , 39.        , 55.        , 31.        ,
       34.        , 15.        , 28.        ,  8.        , 19.        ,
       40.        , 66.        , 42.        , 21.        , 18.        ,
        3.        ,  7.        , 49.        , 29.        , 65.        ,
       28.5       ,  5.        , 11.        , 45.        , 17.        ,
       32.        , 16.        , 25.        ,  0.83      , 30.        ,
       33.        , 23.        , 24.        , 46.        , 59.        ,
       71.        , 37.        , 47.        , 14.5       , 70.5       ,
       32.5       , 12.        ,  9.        , 36.5       , 51.        ,
       55.5       , 40.5       , 44.        ,  1.        , 61.        ,
       56.        , 50.        , 36.        , 45.5       , 20.5       ,
       62.        , 41.        , 52.        , 63.        , 23.5 

In [38]:

age_mean = df2['age'].mean()

df2['age'] = df2['age'].fillna(age_mean)

print(f"Dataset ka mean age tha: {age_mean}")
print(df2['age'].isnull().sum())

Dataset ka mean age tha: 29.69911764705882
0


In [39]:
df2

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.000000,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.000000,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.000000,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.000000,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.000000,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.000000,0,0,13.0000,S,Second,man,True,NaN,Southampton,no,True
887,1,1,female,19.000000,0,0,30.0000,S,First,woman,False,B,Southampton,yes,True
888,0,3,female,29.699118,1,2,23.4500,S,Third,woman,False,NaN,Southampton,no,False
889,1,1,male,26.000000,0,0,30.0000,C,First,man,True,C,Cherbourg,yes,True


In [40]:
df2.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [41]:
df3 = df2

In [42]:
df3.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [43]:
df3 = df3.drop(columns = ['deck'])

In [44]:
df3.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [45]:
df4 = df3

In [46]:
embark_town_mode = df4['embark_town'].mode()

df4['embark_town'] = df4['embark_town'].fillna((embark_town_mode)[0], inplace = True)

print(f"Dataset embark_town ka mode tha: {embark_town_mode}")
print(df4['embark_town'].isnull().sum())


embarked_mode = df4['embarked'].mode()

df4['embarked'] = df4['embarked'].fillna((embarked_mode)[0], inplace = True)

print(f"Dataset embarked ka mode tha: {embarked_mode}")
print(df4['embarked'].isnull().sum())

Dataset embark_town ka mode tha: 0    Southampton
Name: embark_town, dtype: str
0
Dataset embarked ka mode tha: 0    S
Name: embarked, dtype: str
0


C:\Users\hafib\AppData\Local\Temp\ipykernel_1208\4268497630.py:3: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df4['embark_town'] = df4['embark_town'].fillna((embark_town_mode)[0], inplace = True)
C:\Users\hafib\AppData\Local\Temp\ipykernel_1208\4268497630.py:11: ChainedAssignmentError: A value is being set on a copy of a DataFrame 

In [47]:
df4.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [48]:
print('mean = ', df['age'].mean())
print('median = ', df['age'].median())
print('mode = ', df['age'].mode())

mean =  29.69911764705882
median =  29.69911764705882
mode =  0    29.699118
Name: age, dtype: float64


# Steps: Data Wrangling (EDA)
1. Importing libraries.
2. Importing datasets.
3. Explore your datasets.
   1. Information
   2. Data types
   3. Missing values
   4. Take sense of your data
4. Understanding the variables.
5. Relationship between the variables.(heatmap, correlation, pairplot)
6. Brainstorming.
   1. Normalize
   2. Removing outliers 
7. Tidy data, clean data  
8. ready for statistical analysis
9. ready for prediction
10. ready for machine learning
11. ready of DL
12. 

In [ ]:

# Assignment : Normalize krne ke techniques konsi hai
# outliers kia hoty hai or kese kam krte hai?

# Assignment:
how to deal with missing values in EDA?

1. categorical / object

2. numeric (float, int)

3. Boolean



---

## Step 1: Pehle check karo missing values hain kitni?

Sabse pehle dataset load karke missing values ka status check karte hain.

```python
import pandas as pd
import numpy as np

# Titanic dataset load karein (agar aapke paas local file hai)
df = pd.read_csv('titanic.csv')

# Missing values ka percentage dekhne ke liye
print((df.isnull().sum() / len(df)) * 100)

```

**Output mein aapko generally yeh scene dikhega:**

* `Age`: ~20% missing values.
* `Cabin`: ~77% missing values (Bohot zyada!).
* `Embarked`: ~0.2% missing values (Sirf 2 rows missing).

---

## Step 2: Har datatype aur column ko handle karna

### 1. Numeric Column: `Age` (~20% Missing)

`Age` ek numeric column hai. Kyunki missing data sirf 20% hai, toh hum rows delete nahi karenge. Hum isse **Median** se fill karenge kyunki age mein kuch bache aur kuch bohot boorhe log bhi the (outliers).

* **Aasan Approach (Overall Median):** Poore dataset ka jo median age hai, usse fill kar do.

```python
age_median = df['Age'].median()
df['Age'].fillna(age_median, inplace=True)

```

* **Smart Approach (Advanced EDA):** Sirf blind median lagane ke bajaye, logon ke Naam ke sath lage Title (`Mr.`, `Mrs.`, `Miss`, `Master`) ke hisab se group karke median nikalo. Kyunki `Master` (bache) ki median age kam hogi aur `Mr.` ki zyada.

```python
# Naam se Title nikalne ka shortcut
df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)

# Har Title ke hisab se Age ka median fill karna
df['Age'] = df.groupby('Title')['Age'].transform(lambda x: x.fillna(x.median()))

```

---

### 2. Categorical Column (High Missing): `Cabin` (~77% Missing)

`Cabin` ek object/categorical column hai. Ismein **77% data missing** hai. Agar hum isko mode se fill karenge toh data corrupt ho jayega.

* **Best Approach:** Is missing status ko hi ek nayi category bana do (`Unknown` ya `M`), kyunki ship par jinka cabin number nahi likha tha, high chances hain ki wo lower class (Third Class) ke musafir the. Yani missing hona bhi ek statistical information hai!

```python
df['Cabin'].fillna('Unknown', inplace=True)

```

---

### 3. Categorical Column (Low Missing): `Embarked` (0.2% Missing)

`Embarked` (jahan se musafir ship par sawaar hue) mein sirf **2 rows missing** hain.

* **Best Approach:** Kyunki yeh categorical data hai aur missing values bohot hi kam hain, toh hum **Mode** (jo city sabse zyada baar repeat hui hai) use karenge. Titanic par sabse zyada log 'S' (Southampton) se charhe the.

```python
# Sabse zyada repeat hone wali city (Mode)
mode_embarked = df['Embarked'].mode()[0]

# Fill kar do
df['Embarked'].fillna(mode_embarked, inplace=True)

```

---

## Step 3: Final Verification

Sab kuch handle karne ke baad, dubara check karo ki koi missing value bachi toh nahi:

```python
print(df.isnull().sum())

```

Ab aapka output har column ke samne **0** dikhayega, aur aapka Titanic data machine learning model ke liye bilkul ready hai!